mount/mount.go

```diff
@@ -159,21 +159,21 @@ func (h *querySet) next() (query.Result, bool) {
 // /aoe/      not matching
 func (d *Datastore) lookupAll(key ds.Key) (dst []ds.Datastore, mountpoint, rest []ds.Key) {
 	for _, m := range d.mounts {
-		p := m.Prefix.String()
-		if len(p) > 1 {
-			p = p + "/"
-		}
-
-		if strings.HasPrefix(p, key.String()) {
+		if m.Prefix.IsDescendantOf(key) {
 			dst = append(dst, m.Datastore)
 			mountpoint = append(mountpoint, m.Prefix)
 			rest = append(rest, ds.NewKey("/"))
-		} else if strings.HasPrefix(key.String(), p) {
+		} else if m.Prefix.Equal(key) || m.Prefix.IsAncestorOf(key) {
 			r := strings.TrimPrefix(key.String(), m.Prefix.String())
 
 			dst = append(dst, m.Datastore)
 			mountpoint = append(mountpoint, m.Prefix)
 			rest = append(rest, ds.NewKey(r))
+
+			// We've found an ancestor (or equal) key. We might have
+			// more general datastores, but they won't contain keys
+			// with this prefix so there's no point in searching them.
+			break
```

Comment:
This is https://github.com/ipfs/go-datastore/pull/140#discussion_r352838087. Thinking back on this, I think it was a bug.

 Given the mounts `["/foo", "/"]`, `lookupAll("/foo/bar")` would return _both_ (there's now a test for this). However, it should have only returned "/foo" because no keys under "/foo/bar" would end up in "/".

Reply:
Note: this may have been impacting the performance of GC.

Reply:
We are, but that should be fine. That is, we're querying for blocks in /blocks, not for blocks in /. As of this change, `lookupAll("/blocks")`, should return only the datastore mounted at /blocks.